tests/e2e/tests/pilot/pod_churn_test.go

```diff
@@ -0,0 +1,281 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package pilot
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/test/application/echo"
+	"istio.io/istio/pkg/test/application/echo/proto"
+	"istio.io/istio/pkg/test/kube"
+)
+
+const (
+	churnAppName         = "churn"
+	churnAppSelector     = "app=" + churnAppName
+	churnAppReplicas     = 10
+	churnAppMinReplicas  = 5
+	churnTrafficDuration = 1 * time.Minute
+
+	// churnTrafficThreads is the number of threads that will concurrently send traffic to the service being
+	// churned. This is set to Envoy's default circuit breaker parameter max_retry_attempts, which
+	// is the maximum number of parallel retries that Envoy will allow to the upstream cluster. If we
+	// go beyond that threshold, the circuit breaker itself will cause 503s.
+	churnTrafficThreads = 3
+)
+
+// TestPodChurn creates a replicated app and periodically brings down one pod. Traffic is sent during the pod churn
+// to ensure that no 503s are received by the application. This verifies that Envoy is properly retrying disconnected
+// endpoints.
```

Comment:
Yeah, you can definitely do this with 2 pods, where you periodically kill one

Reply:
My concern with only having one pod available is the race of when Pilot informs the Envoys of the available Pod. I'm concerned about a case where Envoy's current endpoints list contains only bad endpoints.

To avoid ever getting into that situation, I've set the minimum number of replicas that are in the ready state to `2`. I've also bumped up the total replicas to `5` to allow churning between 2 and 5.